tests/block_atx/add_node/test_blocks_add_node.py

```diff
@@ -0,0 +1,101 @@
+import time
+from pytest_testconfig import config as testconfig
+
+from tests import queries as q
+from tests.test_bs import setup_network, add_curl, setup_bootstrap, start_poet, setup_clients, wait_genesis, get_conf
+from tests.test_bs import add_multi_clients
+from tests.utils import validate_blocks_per_nodes, get_pod_id
+
+
+# epoch i:
+# start with x miners
+# wait 2 epochs
+#
+# epoch i+2:
+# add a miner
+#
+# epoch i+3
+# validate total miner generated Tavg/x (floored) in i+2
+#
+# epoch i+4:
+# validate total miner generated Tavg/x (floored) in i+3
+# validate number of ATXs = num_miner + 1 in i+3
+#
+# epoch i+5
+# validate total miner generated Tavg/x (floored) in i+4
+#
+# epoch i+6
+# validate total miner generated Tavg/x+1 (floored) in i+5
+def test_add_node_validate_atx(init_session, setup_network):
+    curr_epoch = 0
+    epochs_to_sleep = 2
+    layer_duration = int(testconfig['client']['args']['layer-duration-sec'])
+    layers_per_epoch = int(testconfig['client']['args']['layers-per-epoch'])
+    layer_avg_size = int(testconfig['client']['args']['layer-average-size'])
+    num_miners = int(testconfig['client']['replicas']) + 1  # add 1 for bs node
+
+    print(f"\nlayer duration={layer_duration}, layers per epoch={layers_per_epoch}, layer avg size={layer_avg_size}")
+    # wait for 2 epochs
+    last_layer = epochs_to_sleep * layers_per_epoch
+    print(f"wait until second epoch to layer {last_layer}")
+    _ = q.wait_for_latest_layer(init_session, last_layer, layers_per_epoch)
+
+    # ========================== epoch i+2 ==========================
+    curr_epoch += epochs_to_sleep
+    print("\n\n-------- current epoch", curr_epoch, "--------")
+    print("adding a new miner")
+    bs_info = setup_network.bootstrap.pods[0]
+    cspec = get_conf(bs_info, testconfig['client'])
+    new_pod_name = add_multi_clients(init_session, cspec, 1)[0]
+
+    # wait for next epoch
+    last_layer = layers_per_epoch * (curr_epoch + 1)
+    print(f"wait until next epoch to layer {last_layer}")
+    _ = q.wait_for_latest_layer(init_session, last_layer, layers_per_epoch)
+
+    # ========================== epoch i+3 ==========================
+    curr_epoch += 1
+    print("\n\n-------- current epoch", curr_epoch, "--------")
+    block_map, _ = q.get_blocks_per_node_and_layer(init_session)
+    print(f"-------- validating blocks per nodes up to layer {last_layer} --------")
+    # sometimes block_map contains the newly created node with value of 0
+    # this is an unwanted behavior, until fixed exclude new created node
```

Comment:
don't understand this comment, what happens here and why not fix it?

Reply:
removing this comment

Reply:
fixing get_blocks_per_node_and_layer

Reply:
nothing to fix, this behaviour happens because we query for all "block created" messages and not only for a specific epoch, which means we'll have the node's key in our result yet in some epochs if the node didn't exist we will get '0' as a result